### PR TITLE
Pass additional information when calculating HTTP retry delay

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

In `azure-core` a new API is being added to `RetryStrategy` which takes `RequestRetryCondition`, allowing for additional context about the request failure to be passed when calculating the retry delay.

In ClientCore the `Function` taking `HttpHeaders` and retrying a duration delay has been changed to take `HttpRetryCondition`, for the same reason as mentioned above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java [hunk 1]
```diff
@@ -310,7 +311,14 @@ private void logRetry(LoggingEvent log, int tryCount, Duration delayDuration, Ru
         }
     }
 
-    private Duration calculateRetryDelay(int retryAttempts) {
+    private Duration calculateRetryDelay(HttpRetryCondition retryCondition) {
+        if (delayFromRetryCondition != null) {
+            Duration delay = delayFromRetryCondition.apply(retryCondition);
+            if (delay != null) {
+                return delay;
+            }
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java:300]: This seems like an invalid `delayFromRetryCondition` function if it returns `null` and should throw instead of falling back to other types delay durations.

**Review by srnagar — APPROVED** [review]

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java:300]: Good catch, will change this so if that returns null it will throw

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java [hunk 2]
```diff
@@ -140,19 +143,18 @@ public HttpPipelinePosition getPipelinePosition() {
     /*
      * Determines the delay duration that should be waited before retrying using the well-known retry headers.
      */
-    private Duration getWellKnownRetryDelay(HttpHeaders responseHeaders, int tryCount,
-        Supplier<OffsetDateTime> nowSupplier) {
-        Duration retryDelay = getRetryAfterFromHeaders(responseHeaders, nowSupplier);
+    private Duration getWellKnownRetryDelay(HttpRetryCondition retryCondition, Supplier<OffsetDateTime> nowSupplier) {
+        Duration retryDelay = getRetryAfterFromHeaders(retryCondition.getResponse().getHeaders(), nowSupplier);
```
Comment (g2vinay) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java:148]: will a response always be there ?
can get Response return null ? and throw NPE here ?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java [hunk 3]
```diff
@@ -310,7 +311,17 @@ private void logRetry(LoggingEvent log, int tryCount, Duration delayDuration, Ru
...
     }
 
-    private Duration calculateRetryDelay(int retryAttempts) {
+    private Duration calculateRetryDelay(HttpRetryCondition retryCondition) {
+        if (delayFromRetryCondition != null) {
+            Duration delay = delayFromRetryCondition.apply(retryCondition);
+            if (delay != null) {
+                return delay;
+            } else {
+                throw LOGGER.throwableAtWarning()
```
Comment (g2vinay) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java]: Should this be documented as breaking behavior in changelog too ?

File: sdk/core/azure-core/src/test/java/com/azure/core/http/policy/RetryPolicyTests.java [hunk 4]
```diff
@@ -440,16 +443,17 @@ public Mono<HttpResponse> send(HttpRequest request) {
...
     @MethodSource("getWellKnownRetryDelaySupplier")
     public void getWellKnownRetryDelay(HttpHeaders responseHeaders, RetryStrategy retryStrategy, Duration expected) {
-        assertEquals(expected,
-            RetryPolicy.getWellKnownRetryDelay(responseHeaders, 1, retryStrategy, OffsetDateTime::now));
+        HttpResponse response = new MockHttpResponse(null, 0, responseHeaders);
+        assertEquals(expected, RetryPolicy.getWellKnownRetryDelay(response, 1, retryStrategy, OffsetDateTime::now));
     }
 
     @Test
     public void retryAfterDateTime() {
```
Comment (g2vinay) [inline sdk/core/azure-core/src/test/java/com/azure/core/http/policy/RetryPolicyTests.java:451]: missing test case to validate exception throw when delay is null.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java:148]: It shouldn't be null, but if it were this code path would've thrown an NPE before this change.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/HttpRetryPolicy.java]: I'm actually going to revert this as we've always used a fallback behavior here and I think we should continue that.

**Review by g2vinay — APPROVED** [review]
